source/client/client_worker_impl.cc

```diff
@@ -17,34 +22,55 @@ ClientWorkerImpl::ClientWorkerImpl(Envoy::Api::Api& api, Envoy::ThreadLocal::Ins
                                    Envoy::Stats::Store& store, const int worker_number,
                                    const Envoy::MonotonicTime starting_time,
                                    Envoy::Tracing::HttpTracerPtr& http_tracer)
-    : WorkerImpl(api, tls, store), worker_scope_(store_.createScope("cluster.")),
+    : WorkerImpl(api, tls, store), termination_predicate_factory_(termination_predicate_factory),
+      sequencer_factory_(sequencer_factory), worker_scope_(store_.createScope("cluster.")),
       worker_number_scope_(worker_scope_->createScope(fmt::format("{}.", worker_number))),
       worker_number_(worker_number), starting_time_(starting_time), http_tracer_(http_tracer),
       request_generator_(request_generator_factory.create()),
       benchmark_client_(benchmark_client_factory.create(
           api, *dispatcher_, *worker_number_scope_, cluster_manager, http_tracer_,
-          fmt::format("{}", worker_number), *request_generator_)),
-      termination_predicate_(
-          termination_predicate_factory.create(time_source_, *worker_number_scope_, starting_time)),
-      sequencer_(sequencer_factory.create(time_source_, *dispatcher_, starting_time,
-                                          *benchmark_client_, *termination_predicate_,
-                                          *worker_number_scope_)) {}
+          fmt::format("{}", worker_number), *request_generator_)) {}
 
 void ClientWorkerImpl::simpleWarmup() {
-  ENVOY_LOG(debug, "> worker {}: warmup start.", worker_number_);
-  if (benchmark_client_->tryStartRequest([this](bool, bool) { dispatcher_->exit(); })) {
-    dispatcher_->run(Envoy::Event::Dispatcher::RunType::RunUntilExit);
-  } else {
-    ENVOY_LOG(warn, "> worker {}: failed to initiate warmup request.", worker_number_);
-  }
-  ENVOY_LOG(debug, "> worker {}: warmup done.", worker_number_);
+  // We add a short warmup phase, which ends when either the first successful response
+  // is observed or two seconds have passed, whichever comes first. These warmup conditions are
+  // registered on top of the original configured predicates.
+  // TODO(oschaaf): allow configuration of this phase once the ramping rate limiters land,
+  // or a generic configuration of multiple phases (id, duration, termination predicates, rate
```

Comment:
(for the future) Instinctively I lean towards generic configuration of multiple phases, rather than built-in phase types, but I haven't given it as much thought as you did. Generic configuration for phases will allow users to utilize phases in ways that we didn't account for.

On the other hand the warmup phase is very likely going to be fairly common. Maybe we could come up with providing some sort of templating for command line / service arguments that we could ship with the final version of phases. I.e. instead of building the warmup phase (or others) into the code we could bake it into config templates. WDYT?

Reply:
My instinct is the same; however because of the intermediate step performed here we're applying the generic phases concept in a very specific way to ensure we're doing a no-op. Hopefully we won't have hard coded build-in phases when we are done; I love the config templates idea.

Reply:
Backed some of this out: we retain the old warmup call, and have a single `phase` for now to minimize changes that won't have future value.